### PR TITLE
Supports Object and Array for `extraHeaders` option in loadURL

### PR DESCRIPTION
**Pull request by karupanerura**
I need it:
```es6
webContents.loadUrl('https://www.example.com', {
  extraHeaders: {
    'Accept-Encoding': 'gzip;q=0,deflate;q=0'
  }
});
```

### CODE REVIEW

File: lib/browser/api/navigation-controller.js [hunk 1]
```diff
@@ -58,10 +58,29 @@ var NavigationController = (function () {
   }
 
   NavigationController.prototype.loadURL = function (url, options) {
-    if (options == null) {
+    if (!options) {
```
Comment (kevinsawicki) [inline lib/browser/api/navigation-controller.js]: Was there a reason for changing this `== null` check to a `!` instead?

Comment (karupanerura) [inline lib/browser/api/navigation-controller.js]: oh! sorry, it's my mistake. I'll revert it soon.

File: lib/browser/api/navigation-controller.js [hunk 2]
```diff
@@ -62,6 +62,25 @@ var NavigationController = (function () {
       options = {}
     }
     this.pendingIndex = -1
+    if (options.hasOwnProperty('extraHeaders')) {
+      let extraHeaders = options['extraHeaders']
+      if (typeof extraHeaders === 'object') {
```
Comment (kevinsawicki) [inline lib/browser/api/navigation-controller.js]: The `typeof` value for arrays is `'object'` so I'm not sure this is working correctly for the array case.

File: lib/browser/api/navigation-controller.js [hunk 3]
```diff
@@ -62,6 +62,25 @@ var NavigationController = (function () {
...
+            newExtraHeaders.push([key, extraHeaders[key]])
+          }
+        }
+        extraHeaders = newExtraHeaders
+      }
+      if (Array.isArray(extraHeaders)) {
+        // Escape header values for security:
+        //   replace LWS with a single SP
+        //   remove CR and LF since the char is invalid here
+        extraHeaders = extraHeaders.map(pair => pair.map(v => v.replace(/\015\012[\040\011]/g, ' ').replace(/[\015\012]/g, '')).join(': ')).join('\n')
```
Comment (kevinsawicki) [inline lib/browser/api/navigation-controller.js]: This is a lot for single line, would you mind reformatting this to break it up across multiple lines to make it easier to see how the object is being transformed.

**Comment by kevinsawicki** [comment]
Left a few minor comments, it would be great to get test coverage of these new options to ensure things are working as expected and to have examples of how this can be used with the new supported parameter types.

File: lib/browser/api/navigation-controller.js [hunk 4]
```diff
@@ -62,6 +62,25 @@ var NavigationController = (function () {
...
+      }
+      options['extraHeaders'] = extraHeaders
```
Comment (kevinsawicki) [inline lib/browser/api/navigation-controller.js]: Rather that writing back to this object, I think it should be copied earlier on in this method, that way the caller isn't surprised that the object they passed in was changed on them.

**Comment by karupanerura** [comment]
Thank you for your review. I'll fix it on next week.

**Comment by kevinsawicki** [comment]
> Thank you for your review. I'll check and fix it on next week.

Sounds good, going to close this out for now then, please comment when you push new changes up and we can reopen it, thanks.

**Pull request closed without merging**